### PR TITLE
Calculator REPL parses and evaluates infix arithmetic expressions

### PR DESCRIPTION
**Pull request by bradley-gersh**
The REPL calculator should now process and evaluate arithmetic expressions using parentheses and the usual operators `+ - * /`, including nested expressions. A representation of the parse tree in prefix notation (idea from [here](http://craftinginterpreters.com/representing-code.html)) should also display if you start your expression with `?`, like this:

```
> 4 * (4 - 6)
= -8.0

> ?4 * (4 - 6)
--> parse tree: (* 4.0 (- 4.0 6.0))
= -8.0
```

Some questions that came up:

~~1. Tokens are implemented as case classes, but really they're all constants except the `NUMBER(value)` tokens. Is this OK style? Originally I used enums: I had an `enum TokenType` that contained the constant token names and used `case class Token(type: TokenType, content: String)` to create each token, but that quickly got verbose, especially when writing tests.~~ Resolved by using case objects.

2. There are a lot of nested `if`/`else` statements, usually to guard against bad input to a function. (See the `parseRoot` method in `Parser.scala` for an example.) Is this idiomatic? In an imperative language I would break early with `return` statements, but here I was avoiding using `return` to follow the idea that everything is an expression. The nested conditionals seem excessive though.

3. Scaladoc comments don't seem to work smoothly with top-level functions. My first attempt at the Evaluator looked something like this:

```scala
// first try

/**
 * Evaluates a given parse tree for the return value
 * 
 * @param  tree  root node of the parse tree
 * @return       Try of the final numerical value of the parse tree
 */
val Evaluator = (tree: ParseNode) => {
  val value = Try(evaluate(tree))
  // ...
}
```

The IDE complained that `@param` couldn't go there (since `Evaluator` is a class, not a method), so I had to write the `.apply` method explicitly:

```scala
// current version

/**
 * Evaluates a given parse tree for the return value
 */
object Evaluator extends Function[ParseNode, Try[Double]] {
  /**
   * Public access to Evaluator function
   * 
   * @param  tree  root node of the parse tree
   * @return       Try of the final numerical value of the parse tree
   */
  def apply(tree: ParseNode): Try[Double] = {
    val value = Try(evaluate(tree))
    // ...
  }
}
```

Is there a way to keep the top-level function and describe it using Scaladoc comments?

Thanks for any input!

### CODE REVIEW

File: src/main/scala/it/scalalearn/calculator/tokenTypes.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+sealed abstract class Token(val string: String)
+
+case class LPAREN() extends Token(string ="(")
+case class RPAREN() extends Token(string = ")")
+case class DOT() extends Token(string = ".")
+case class PLUS() extends Token(string = "+")
+case class DASH() extends Token(string = "-")
+case class STAR() extends Token(string = "*")
+case class SLASH() extends Token(string = "/")
+case class WS() extends Token(string = " ")
```
[the remark this answers is not in the document]
Comment (bradley-gersh) [reply src/main/scala/it/scalalearn/calculator/tokenTypes.scala]: Thanks! That makes the tokens much cleaner.

File: src/main/scala/it/scalalearn/calculator/Calculator.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ */
+object Calculator {
+
+  /**
+   * Entry point: Print banner and run REPL
+   */
+  def main(args: Array[String]): Unit = {
+    println("ScalaCalc v0.1\n\n"
+      + "Enter an empty line or ctrl-D to quit (ctrl-Z + enter on Windows).\n\n"
+      + "Begin input with a single ? character to view the parse tree for the expression.\n")
```
Comment (ltronky) [inline src/main/scala/it/scalalearn/calculator/Calculator.scala:18]: Nice feature in scala you have trippleqoote nearly like in python

```suggestion
    println("""
      | ScalaCalc v0.1
      |
      | Enter an empty line or ctrl-D to quit (ctrl-Z + enter on Windows).
      |
      | Begin input with a single ? character to view the parse tree for the expression.
      |""".stripMargin)
```

**Comment by ltronky** [comment]
I see you are using exceptions to handle "exceptional cases". As for IFs you would like to avoid `return` statements, this is the same with `throw`. They are jumps that don't respect the semantics of the function.

In this case, we have types like `Either` or `Try`, that models something that could be good or bad / right or left / success or failure.
In this case you could define your error/exception hierarchy and use it as a return value in a function, instead of throwing it.

File: src/main/scala/it/scalalearn/calculator/Parser.scala [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+   */
+  private def parseSign(tokens: List[Token], parenLevel: List[Token]): (List[Token], ParseNode, List[Token]) = {
+    if (tokens.isEmpty) throw new ParserException("expression terminated where a value was expected")
+
+    val t = tokens.head
+    if (t == DASH) {
+      val (remainingTokens, number, newParenLevel) = parseSign(tokens.tail, parenLevel)
+      (remainingTokens, SignNode(t, number), newParenLevel)
+    }
+    else parseNumber(tokens, parenLevel)
```
Comment (ltronky) [inline src/main/scala/it/scalalearn/calculator/Parser.scala:160]: Pattern match would help here in readability

```suggestion
    tokens match {
      case Nil => throw new ParserException("expression terminated where a value was expected")
      case DASH :: rest =>
              val (remainingTokens, number, newParenLevel) = parseSign(rest, parenLevel)
              (remainingTokens, SignNode(t, number), newParenLevel)
      case _ => parseNumber(tokens, parenLevel)
```

File: src/main/scala/it/scalalearn/calculator/Calculator.scala (hunk 2 above)
Comment (bradley-gersh) [reply src/main/scala/it/scalalearn/calculator/Calculator.scala:18]: Thanks, I didn't know about `.stripMargin`.

File: src/main/scala/it/scalalearn/calculator/Parser.scala (hunk 3 above)
Comment (bradley-gersh) [reply src/main/scala/it/scalalearn/calculator/Parser.scala:160]: That `DASH :: rest` pattern is very useful. I'll see if I can use the same idea elsewhere in that file.

File: src/main/scala/it/scalalearn/calculator/Evaluator.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+          if (numerator == 0.0) throw new ParserException("indeterminate form 0/0 obtained")
+          else throw new ParserException("division by zero")
+        } else numerator / denominator
+      }
+      else throw new ParserException(s"improper operation ${op.string} where multiplication or division was expected")
+
+    case TermNode(op, expr1, expr2) =>
+      if (op == PLUS) evaluate(expr1) + evaluate(expr2)
+      else if (op == DASH) evaluate(expr1) - evaluate(expr2)
+      else throw new ParserException(s"improper operation ${op.string} where addition or subtraction was expected")
```
[the remark this answers is not in the document]
Comment (bradley-gersh) [reply src/main/scala/it/scalalearn/calculator/Evaluator.scala:69]: Looks good, will revisit the other cases in the same way.

**Comment by bradley-gersh** [comment]
> I see you are using exceptions to handle "exceptional cases". As for IFs you would like to avoid `return` statements, this is the same with `throw`. They are jumps that don't respect the semantics of the function.
> 
> In this case, we have types like `Either` or `Try`, that models something that could be good or bad / right or left / success or failure. In this case you could define your error/exception hierarchy and use it as a return value in a function, instead of throwing it.

I see your point. I used `Try`s at higher levels of abstraction, but I guess they should be used throughout the project. I will add them in.

File: src/main/scala/it/scalalearn/calculator/treeTypes.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+sealed trait ParseNode() {
+   def isEmpty = false
+}
+
+sealed trait FuncNode(val op: Token, val exprs: ParseNode*) extends ParseNode
+
+case class EmptyNode() extends ParseNode {
+   override def isEmpty = true
+}
```
[the remark this answers is not in the document]
Comment (bradley-gersh) [reply src/main/scala/it/scalalearn/calculator/treeTypes.scala]: Yes, I caught that one this morning while try to refactor all the exception handling into Either. Thanks!

**Pull request merged**